src/main/scala/com/fulcrumgenomics/umi/UmiConsensusCaller.scala

```diff
@@ -0,0 +1,277 @@
+    if (ids.map(_.length+1).sum <= 200) ids.mkString("|")
+    else Integer.toHexString(new Murmur3(1).hashUnencodedChars(ids.mkString("|")))
+  }
+
+  /**
+    * Constructs an output header with a single read group for the a BAM.
+    */
+  def outputHeader(in: SAMFileHeader, readGroupId: String, sortOrder: Option[SortOrder] = None): SAMFileHeader = {
+    val oldRgs = in.getReadGroups.toSeq
+    def collapse(f: SAMReadGroupRecord => String): String = oldRgs.map(f).filter(_ != null).toSet.toList match {
+      case Nil      => null
+      case x :: Nil => x
+      case xs       => xs.mkString(",")
+    }
+
+    val rg = new SAMReadGroupRecord(readGroupId)
+    rg.setDescription(s"Consensus reads generated from ${oldRgs.size} input read groups.")
+    rg.setLibrary         (collapse(_.getLibrary))
+    rg.setSample          (collapse(_.getSample))
+    rg.setPlatform        (collapse(_.getPlatform))
+    rg.setPlatformUnit    (collapse(_.getPlatformUnit))
+    rg.setSequencingCenter(collapse(_.getSequencingCenter))
+
+    val outHeader = new SAMFileHeader
+    outHeader.addReadGroup(rg)
+    outHeader.setSortOrder(sortOrder.getOrElse(SortOrder.unsorted))
+    outHeader.setGroupOrder(GroupOrder.query)
+    outHeader
+  }
+}
+
+
+/**
+  * A trait that can be mixed in by any consensus caller that works at the read level,
+  * mapping incoming SAMRecords into consensus SAMRecords.
+  *
+  * @tparam C Internally, the type of lightweight conesnsus read that is used prior to
+  *           rebuilding SAMRecords.
+  */
+trait UmiConsensusCaller[C <: ConsensusRead] {
+  import com.fulcrumgenomics.umi.UmiConsensusCaller.ReadType._
+
+  // vars to track how many reads meet various fates
+  private var _totalReads: Long = 0
+  private var _totalFiltered: Long = 0
+  private var _filteredForMinorityAlignment: Long = 0
+
+  /** Returns the total number of input reads examined by the consensus caller so far. */
+  def totalReads: Long = _totalReads
+
+  /**
+    * Returns the number of raw reads filtered out due to there being insufficient reads present
+    * to build the necessary set of consensus reads.
+    */
+  def readsFilteredInsufficientSupport: Long = this._totalFiltered - this._filteredForMinorityAlignment
+
+  /** Returns the number of raw reads filtered out because their alignment disagreed with the majority alignment of
+    * all raw reads for the same source molecule
+    */
+  def readsFilteredMinorityAlignment: Long = _filteredForMinorityAlignment
+
+  /** Records that the supplied records were rejected, and not used to build a consensus read. */
+  protected def rejectRecords(recs: Traversable[SAMRecord]) : Unit = {
+    this._totalFiltered += recs.size
+  }
+
+  /** A RG.ID to apply to all generated reads. */
+  protected def readGroupId: String
+
+  /** A prefix to use on all read names.  If None then a suitable prefix will be synthesized. */
+  protected val readNamePrefix: String
```

Comment:
It's an `Option[String]` on the command line programs, but a `String` here, because I wanted to do away with the need to feed the _input_ `SAMFileHeader` into the UMI caller.  There's a utility method in the companion object to make a prefix from the input header, so that's still available.  

I would only want to make it an option here if it's truly optional, and when absent we don't use a prefix at all.  Is that what you're suggesting?  That would make valid BAM files that would be nearly impossible to merge with other consensus BAM files later.